### PR TITLE
Add chart timer uptime data

### PR DESCRIPTION
**Pull request by Smanmos**
Add timer uptime data to the GUI. This can be toggled by shift+right click, disabled by default. 

If updating, data/db needs to be deleted. 

To accommodate this change, the remote cache is no longer used. However, optimisations have been made to calculate the note count of types faster.

### CODE REVIEW

File: src/gui/viewmodels/song.py [hunk 1]
```diff
@@ -56,9 +67,12 @@ def __init__(self, main):
         self.widget.setDragEnabled(True)
         self.widget.setSelectionMode(QAbstractItemView.SingleSelection)
         self.widget.setSelectionBehavior(QAbstractItemView.SelectRows)
-        self.widget.setToolTip("Right click to show percentage.")
+        self.widget.setToolTip("Right click to toggle note types between number and percentage.\n"
+                               "Shift-right click to toggle timer percentages")
```
Comment (dlresearchsociety) [inline src/gui/viewmodels/song.py:71]: Missing period: `Shift-right click to toggle timer percentages.`

File: src/network/chart_cache_updater.py [hunk 2]
```diff
@@ -271,9 +290,17 @@ def _overwrite_song_name(expanded_song_list):
     db.cachedb.commit()
 
 
+timers = [(7, 4.5, 'h'), (9, 6, 'h'), (11, 7.5, 'h'), (12, 7.5, 'm'),
+          (6, 4.5, 'm'), (9, 7.5, 'm'), (11, 9, 'm'), (13, 9, 'h')]
+
+
+def is_active(time, period, duration, last_note):
```
Comment (dlresearchsociety) [inline src/network/chart_cache_updater.py:297]: Please use `interval` and `duration` as the standard terms in code.

File: src/network/chart_cache_updater.py [hunk 3]
```diff
@@ -304,6 +331,22 @@ def update_cache_scores():
...
+            total_notes // 4,
+            total_notes // 2,
+            total_notes * 7 // 10,
+            total_notes * 8 // 10,
+            total_notes * 9 // 10,
+            total_notes + 1
+        ])
+        for timer in timers:
+            live_data['Timer_{}{}'.format(timer[0], timer[2])] = np.repeat(
+                [1., 1.1, 1.2, 1.3, 1.4, 1.5, 1.7, 2.], combo_thresholds[1:] - combo_thresholds[:-1]
```
Comment (dlresearchsociety) [inline src/network/chart_cache_updater.py:348]: Try to not use magic numbers. Combo thresholds and bonuses are defined in `static.live_values`.

File: src/network/chart_cache_updater.py [hunk 4]
```diff
@@ -271,9 +290,17 @@ def _overwrite_song_name(expanded_song_list):
     db.cachedb.commit()
 
 
+timers = [(7, 4.5, 'h'), (9, 6, 'h'), (11, 7.5, 'h'), (12, 7.5, 'm'),
+          (6, 4.5, 'm'), (9, 7.5, 'm'), (11, 9, 'm'), (13, 9, 'h')]
```
Comment (dlresearchsociety) [inline src/network/chart_cache_updater.py:294]: Variable name for final static-ish type should be in ALL_CAPS. Also should move to top of the file.

File: src/network/chart_cache_updater.py [hunk 5]
```diff
@@ -304,6 +331,22 @@ def update_cache_scores():
...
+            total_notes // 2,
+            total_notes * 7 // 10,
+            total_notes * 8 // 10,
+            total_notes * 9 // 10,
+            total_notes + 1
+        ])
+        for timer in timers:
+            live_data['Timer_{}{}'.format(timer[0], timer[2])] = np.repeat(
+                [1., 1.1, 1.2, 1.3, 1.4, 1.5, 1.7, 2.], combo_thresholds[1:] - combo_thresholds[:-1]
+            )[is_active(notes_data['sec'], timer[0], timer[1], notes_data.iloc[-1]['sec'])].sum() / (1.41 * total_notes)
```
Comment (dlresearchsociety) [inline src/network/chart_cache_updater.py:349]: What is `1.41`?

File: src/logic/live.py [hunk 6]
```diff
@@ -37,6 +37,13 @@ def classify_note(row):
         return NoteType.SLIDE
 
 
+def classify_note_vectorized(row):
+    return np.choose(row.type - 3, [
```
Comment (dlresearchsociety) [inline src/logic/live.py:41]: Can run a test on all charts except WITCH to cross check with `classify_note`?

**Review by dlresearchsociety — CHANGES_REQUESTED** [review]

File: src/logic/live.py (hunk 6 above)
Comment (Smanmos) [reply src/logic/live.py:41]: I've added tests for this in [d55101e](https://github.com/deresute-tools/deresute-tools/pull/17/commits/d55101e4bcf453ed416cc9500262b3f77d93af15). The tests passed.

**Comment by Smanmos** [comment]
I've fixed all requested changes

**Comment by dlresearchsociety** [comment]
Merged. Thanks for the contribution.

**Pull request merged**